### PR TITLE
Bias Metrics

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Additions proposed in this pull request:

- Added three bias metrics to the fairness module: `WordEmbeddingAssociationTest`, `EmbeddingCoherenceTest`, and `NaturalLanguageInference`.

### CODE REVIEW

**Review by epwalsh — COMMENTED** [review]
This is really great, I especially love the thorough documentation 💯

I just have two comments.

First: it would be good to have an example of how to use these in practice with a transformer model or something. But that could potentially be in a blog post or GitHub Discussion.

Second: these classes don't have any attributes, so it seems like they should just be functions.

**Comment by ArjunSubramonian** [comment]
> This is really great, I especially love the thorough documentation 💯
> 
> I just have two comments.
> 
> First: it would be good to have an example of how to use these in practice with a transformer model or something. But that could potentially be in a blog post or GitHub Discussion.
> 
> Second: these classes don't have any attributes, so it seems like they should just be functions.

Thanks for the comment! 1) After playing around with the Transformers the past couple of days, I think I'm going to spend next week turning all my metrics and algorithm implementations into wrappers for Transformers and TextFieldEmbedders. For example, to use the bias mitigation algorithms effectively with BERT, you have to finetune with the bias mitigation layer inserted after the base word embedding weights. It would be nice to have a wrapper that handles this + examples of how to use the wrapper. 2) Yup, I agree, I think my initial thought was to make them differentiable, but it doesn't make too much sense to make these metrics differentiable given that they're not particularly smooth.

**Comment by ArjunSubramonian** [comment]
> This is really great, I especially love the thorough documentation 💯
> 
> I just have two comments.
> 
> First: it would be good to have an example of how to use these in practice with a transformer model or something. But that could potentially be in a blog post or GitHub Discussion.
> 
> Second: these classes don't have any attributes, so it seems like they should just be functions.

Sorry, I changed my mind regarding this. I would prefer to keep them as classes right now, in case I want to make them `Registrable`s later on for use in config files.

File: allennlp/fairness/bias_metrics.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+        # Normalize
+        mean_target_embedding1 = torch.nn.functional.normalize(mean_target_embedding1, p=2, dim=-1)
+        mean_target_embedding2 = torch.nn.functional.normalize(mean_target_embedding2, p=2, dim=-1)
+        attribute_embeddings = torch.nn.functional.normalize(attribute_embeddings, p=2, dim=-1)
+
+        # Compute cosine similarities
+        AB_sim_m = torch.matmul(attribute_embeddings, mean_target_embedding1)
+        AB_sim_f = torch.matmul(attribute_embeddings, mean_target_embedding2)
+
+        return self.spearman_correlation(AB_sim_m, AB_sim_f)
```
Comment (AkshitaB) [inline allennlp/fairness/bias_metrics.py:228]: Is there a reason to not use the `scipy.stats` implementation here?

**Review by AkshitaB — COMMENTED** [review]
The code looks fine. How do you envision the use of the bias metrics? For instance, right now, the NLI score is essentially just measuring the accuracy for a specific label, for a single batch of predictions. Is this something that we may want to evaluate over multiple batches?

File: allennlp/fairness/bias_metrics.py (hunk 1 above)
Comment (ArjunSubramonian) [reply allennlp/fairness/bias_metrics.py:228]: I would like to keep everything differentiable if possible.

**Comment by ArjunSubramonian** [comment]
> The code looks fine. How do you envision the use of the bias metrics? For instance, right now, the NLI score is essentially just measuring the accuracy for a specific label, for a single batch of predictions. Is this something that we may want to evaluate over multiple batches?

I'm considering making the bias metrics `Registrable` if I have time, so that they can be specified in the config file to be computed post-training or post-finetuning. I don't personally see these being useful as accumulating metrics, they would be better as validation or test metrics.

**Review by AkshitaB — APPROVED** [review]
LGTM.
